source/fab/metrics.py

```diff
@@ -0,0 +1,214 @@
+
+    """
+    # An example metric is the time taken to preprocess a file; metrics['preprocess c']['my_file.c']
+    metrics: Dict[str, Dict[str, float]] = defaultdict(dict)
+
+    # todo: can we do this better?
+    # we run in a subprocess, so we get a copy of _metric_send_conn before it closes.
+    # when the calling process finally does close it, we'll still have an open copy of it,
+    # so the connection will still be considered open!
+    # therefore we close *OUR* copy of it now.
+    _metric_send_conn.close()
+
+    logger.debug('read_metric: waiting for metrics')
+    num_recorded = 0
+    while True:
+        try:
+            metric = _metric_recv_conn.recv()
+        except EOFError:
+            break
+
+        group, name, value = metric
+        metrics[group][name] = value
+        num_recorded += 1
+
+    logger.debug(f"read_metric: recorded {num_recorded} metrics")
+
+    # Send the collated metrics.
+    # Note: We tried a 2-way pipe so we'd only need one, but it didn't work
+    # so we moved on and created this second pipe for the total metrics.
+    # Might be a simplification in there.
+    _collated_metrics_send_conn.send(metrics)
+    _collated_metrics_send_conn.close()
+
+
+def send_metric(group: str, name: str, value):
+    """
+    Pass a metric to the reader process.
+
+    """
+    _metric_send_conn.send([group, name, value])  # type: ignore
+
+
+def stop_metrics():
+    """
+    Close the metrics pipe and reader process.
+
+    Return the final collection of metrics.
+
+    """
+    global _metric_recv_conn, _metric_send_conn
+    global _collated_metrics_recv_conn, _collated_metrics_send_conn
+    global _metric_recv_process
+
+    # Close the metrics recording pipe.
+    # The metrics recording process will notice and finish,
+    # and send the total metrics to the "collated metrics" pipe, which it then closes.
+    _metric_send_conn.close()  # type: ignore
+    _metric_recv_process.join(1)  # type: ignore
+
+    # pull out the total metrics
+    metrics = _collated_metrics_recv_conn.recv()  # type: ignore
+
+    # set these to none so metrics can be initialised again
+    _metric_recv_conn = _metric_send_conn = _metric_recv_process = _collated_metrics_recv_conn = None
+
+    return metrics
+
+
+def metrics_summary(metrics, output_folder: Path):
+    """
+    Create various summaries fo the metrics, including charts if matplotlib is installed.
+
+    """
+    try:
+        import matplotlib  # type: ignore
+        matplotlib.use('Agg')
+        import matplotlib.pyplot as plt  # type: ignore
+    except ImportError:
+        plt = None
+
+    logger.debug(f'metrics_summary: got metrics for: {metrics.keys()}')
+    output_folder.mkdir(parents=True, exist_ok=True)
+
+    metrics_folder = Path(output_folder)
+    metrics_folder.mkdir(exist_ok=True)
+
+    with open(metrics_folder / 'metrics.json', 'wt') as outfile:
+        json.dump(metrics, outfile, indent='\t')
```

Comment:
Given that this is not Javascript and the ugliness factor, what is the reason for using JSON?

Reply:
:smiley: It's built into Python for a reason. This is a really easy way to save and load a dict with minimal code to maintain. Happy to consider alternatives, of course.